src/ui/components/emote_picker.rs

```diff
@@ -0,0 +1,259 @@
+        let input_suggester = Box::new(|emotes: SharedEmotes, s: String| -> Option<String> {
+            first_similarity_iter(emotes.emotes.borrow().keys(), &s)
+        });
+
+        let input = InputWidget::new(
+            config.clone(),
+            "Emote",
+            Some((emotes.clone(), input_validator)),
+            None,
+            Some((emotes.clone(), input_suggester)),
+        );
+
+        let search_theme = Style::default().fg(Color::Red).add_modifier(Modifier::BOLD);
+
+        Self {
+            config,
+            emotes,
+            input,
+            search_theme,
+            list_state: ListState::default(),
+            filtered_emotes: vec![],
+        }
+    }
+    fn next(&mut self) {
+        let i = match self.list_state.selected() {
+            Some(i) => {
+                if i >= self.filtered_emotes.len() - 1 {
+                    self.filtered_emotes.len() - 1
+                } else {
+                    i + 1
+                }
+            }
+            None => 0,
+        };
+
+        self.list_state.select(Some(i));
+    }
+
+    fn previous(&mut self) {
+        let i = self
+            .list_state
+            .selected()
+            .map_or(0, |i| if i == 0 { 0 } else { i - 1 });
+
+        self.list_state.select(Some(i));
+    }
+
+    fn unselect(&mut self) {
+        self.list_state.select(None);
+    }
+    pub const fn is_focused(&self) -> bool {
+        self.input.is_focused()
+    }
+
+    pub fn toggle_focus(&mut self) {
+        self.input.toggle_focus();
+    }
+}
+
+impl Component for EmotePickerWidget {
+    fn draw(&mut self, f: &mut Frame, area: Option<Rect>) {
+        let r = area.map_or_else(|| centered_rect(60, 60, 20, f.size()), |a| a);
+
+        // Only load the emotes that are actually being displayed, as loading every emote is not really possible.
+        // Some channels can have multiple thousands emotes and decoding all of them takes a while.
+        let max_len = max(
+            self.list_state.selected().unwrap_or_default(),
+            self.list_state.offset(),
+        ) + r.height as usize;
+        let mut items = Vec::with_capacity(max_len);
+        let mut bad_emotes = vec![];
+
+        let current_input = self.input.to_string();
+
+        let cell_size = *self
+            .emotes
+            .cell_size
+            .get()
+            .expect("Terminal cell size should be set when emotes are enabled.");
+
+        let finder = if current_input.is_empty() {
+            None
+        } else {
+            Some(memmem::Finder::new(&current_input))
+        };
+
+        for (name, (filename, zero_width)) in self.emotes.emotes.borrow().iter() {
+            if items.len() >= max_len {
+                break;
+            }
+
+            // Skip emotes that do not contain the current input, if it is not empty.
+            let Some(pos) = finder
+                .as_ref()
+                .map_or_else(|| Some(0), |f| f.find(name.as_bytes()))
```

Comment:
I find that this works fine with the `memmem::Finder` implementation of the search widget. Thoughts?

```suggestion
                .map_or_else(|| Some(0), |f| f.find(name.to_lowercase().as_bytes()))
```

Reply:
![image](https://github.com/Xithrius/twitch-tui/assets/15021300/28eee64c-f50e-440c-94ce-c6483ed47f08)

Reply:
Yes, I did not think about that. I'll implement it tomorrow, thanks.